### ISSUE TITLE
DAN的适配层数

### ISSUE THREAD

**Issue by DianLyu** [jindongwang/transferlearning#55]
王老师您好！近日配合着您的知乎专栏《小王爱迁移》的第四章按DaNN、DDC、DAN的顺序阅读了您的代码实现，注意到DDC、DAN两个实现中的代码都是类似这样的（以下是DAN 中 ResNet.py的代码）：
`class DANNet(nn.Module):

    def __init__(self, num_classes=31):
        super(DANNet, self).__init__()
        self.sharedNet = resnet50(False)
        self.cls_fc = nn.Linear(2048, num_classes)

    def forward(self, source, target):
        loss = 0
        source = self.sharedNet(source)
        if self.training == True:
            target = self.sharedNet(target)
            #loss += mmd.mmd_rbf_accelerate(source, target)
            loss += mmd.mmd_rbf_noaccelerate(source, target)

        source = self.cls_fc(source)
        #target = self.cls_fc(target)

        return source, loss`
我没有使用过Pytorch，但我对这份代码的理解是DAN基于ResNet50，而后跟了一个`self.cls_fc = nn.Linear(2048, num_classes)`层，而forward中的代码代表我们对_一个cls_fc层_的输入进行了mmd适配，但我注意到您在知乎专栏第四章中提到，DAN的改进一是适配了多层（我的理解是需要_对后面多层的输入_都进行MMD约束），二是使用了MKMMD，因此这里是不是并不是与DAN原文完全匹配？一是代码中是_基于ResNet50而非AlexNet_，二是代码这里也是与您的DDC实现一样适配了一层（对_一个cls_fc层_输入进行了约束）。不知我的理解是否正确，还请能够指教！非常感谢！

**Comment by jindongwang** [comment]
你的理解是正确的。DAN的主要贡献就是MK-MMD以及多层适配。但是原作者的实现中确实只适配了一层，这个仍然是个疑问。MK-MMD的实现应该是和原作者一样的，你可以再检查一下。

**Comment by DianLyu** [comment]
嗯嗯，我的主要疑问就是在于不管是您的DAN代码还是THUML Xlearn repo的DAN代码看上去都并没有实现多层适配，原本还以为是我的理解错误，因此非常感谢您的解答！个人觉得您的代码读起来比THUML的实现要易懂很多~再次感谢！

**Comment by jindongwang** [comment]
THUML的DAN实现其实就是DDC加上MK-MMD。我也问过他们为啥没有多层，他们说用一层就好了。。。

**Comment by DianLyu** [comment]
哈哈好吧 那看来其实差距可能就是在MKMMD和MMD之间了。。
